src/Riok.Mapperly/Descriptors/SymbolAccessor.cs

```diff
@@ -0,0 +1,105 @@
+﻿using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Riok.Mapperly.Helpers;
+using Riok.Mapperly.Symbols;
+
+namespace Riok.Mapperly.Descriptors;
+
+public class SymbolAccessor
+{
+    private readonly WellKnownTypes _types;
+    private readonly Dictionary<ISymbol, ImmutableArray<AttributeData>> _cachedAttributes = new(SymbolEqualityComparer.Default);
+    private readonly Dictionary<ITypeSymbol, IReadOnlyCollection<ISymbol>> _allMembers = new(SymbolEqualityComparer.Default);
+    private readonly Dictionary<ITypeSymbol, IReadOnlyCollection<IMappableMember>> _allAccessibleMembers =
+        new(SymbolEqualityComparer.Default);
+
+    public SymbolAccessor(WellKnownTypes types)
+    {
+        _types = types;
+    }
+
+    internal ImmutableArray<AttributeData> GetAttributes(ISymbol symbol)
+    {
+        if (_cachedAttributes.TryGetValue(symbol, out var attributes))
+        {
+            return attributes;
+        }
+
+        attributes = symbol.GetAttributes();
+        _cachedAttributes.Add(symbol, attributes);
+
+        return attributes;
+    }
+
+    internal bool HasAttribute<T>(ISymbol symbol)
+        where T : Attribute
+    {
+        var attributeSymbol = _types.Get(typeof(T));
+        return GetAttributes(symbol).Any(a => SymbolEqualityComparer.Default.Equals(a.AttributeClass, attributeSymbol));
+    }
+
+    internal IEnumerable<IMethodSymbol> GetAllMethods(ITypeSymbol symbol) => GetAllMembers(symbol).OfType<IMethodSymbol>();
+
+    internal IEnumerable<IMethodSymbol> GetAllMethods(ITypeSymbol symbol, string name) =>
+        GetAllMembers(symbol, name).OfType<IMethodSymbol>();
+
+    internal IEnumerable<IPropertySymbol> GetAllProperties(ITypeSymbol symbol, string name) =>
+        GetAllMembers(symbol, name).OfType<IPropertySymbol>();
+
+    internal IEnumerable<IFieldSymbol> GetAllFields(ITypeSymbol symbol) => GetAllMembers(symbol).OfType<IFieldSymbol>();
+
+    internal IReadOnlyCollection<ISymbol> GetAllMembers(ITypeSymbol symbol)
+    {
+        if (_allMembers.TryGetValue(symbol, out var members))
+        {
+            return members;
+        }
+
+        members = GetAllMembersCore(symbol).ToArray();
```

Comment:
Do we really need to collect to array instead of to list? Wouldn't to list be faster here? Same for the accessible mappable members.

Reply:
I'm not sure. Why do you think List is faster?

Reply:
It looks like `members = GetAllMembersCore(symbol)` actually returns `ICollection<ISymbol>`. `ToArray` and `ToList` both have a check for `ICollection` using `CopyTo` under the hood. 
TLDR both should have the same performance (`ToArray` might is faster for an empty ICollection)

Reply:
I think the current solution is fine for now (good performance while still keeping good readability and maintainability).